File: go-broker/internal/physics/integrator_test.go
package physics

import (
	"math"
	"testing"

	"driftpursuit/broker/internal/gameplay"
	pb "driftpursuit/broker/internal/proto/pb"
)

func TestIntegrateVehicleUpdatesLinearAndAngular(t *testing.T) {
	//1.- Construct a vehicle with both velocity components populated.
	state := &pb.VehicleState{
		Position:        &pb.Vector3{X: 1, Y: 2, Z: 3},
		Velocity:        &pb.Vector3{X: 4, Y: -2, Z: 0.5},
		Orientation:     &pb.Orientation{YawDeg: 10, PitchDeg: -5, RollDeg: 0},
		AngularVelocity: &pb.Vector3{X: 20, Y: 30, Z: -10},
	}
	//2.- Advance the state by half a second and verify results.
	IntegrateVehicle(state, 0.5)
	if math.Abs(state.Position.X-3) > 1e-9 {
		t.Fatalf("unexpected X %.2f", state.Position.X)
	}
	if math.Abs(state.Position.Y-1) > 1e-9 {
		t.Fatalf("unexpected Y %.2f", state.Position.Y)
	}
	if math.Abs(state.Position.Z-3.25) > 1e-9 {
		t.Fatalf("unexpected Z %.2f", state.Position.Z)
	}
	if math.Abs(state.Orientation.YawDeg-25) > 1e-9 {
		t.Fatalf("unexpected yaw %.2f", state.Orientation.YawDeg)
	}
	if math.Abs(state.Orientation.PitchDeg-5) > 1e-9 {
		t.Fatalf("unexpected pitch %.2f", state.Orientation.PitchDeg)
	}
	if math.Abs(state.Orientation.RollDeg+5) > 1e-9 {
		t.Fatalf("unexpected roll %.2f", state.Orientation.RollDeg)
	}
}

func TestIntegrateVehicleHandlesInvalidInput(t *testing.T) {
	//1.- Use nil state and zero timestep to ensure safe no-op behaviour.
	IntegrateVehicle(nil, 0.5)
	state := &pb.VehicleState{}
	IntegrateVehicle(state, -1)
	if state.Position != nil || state.Orientation != nil {
		t.Fatalf("integration should not allocate with invalid input")
	}
}

func TestIntegrateVehicleClampsToSkiffStats(t *testing.T) {
	//1.- Build a state with exaggerated linear and angular rates.
	stats := gameplay.SkiffStats()
	state := &pb.VehicleState{
		Position:    &pb.Vector3{X: 0, Y: 0, Z: 0},
		Velocity:    &pb.Vector3{X: stats.MaxSpeedMps * 3, Y: stats.MaxSpeedMps * 2, Z: stats.MaxSpeedMps},
		Orientation: &pb.Orientation{YawDeg: 0, PitchDeg: 0, RollDeg: 0},
		AngularVelocity: &pb.Vector3{
			X: stats.MaxAngularSpeedDegPerSec * 4,
			Y: stats.MaxAngularSpeedDegPerSec * 2,
			Z: 0,
		},
	}
	//2.- Integrate a one second step to exercise the clamp logic.
	IntegrateVehicle(state, 1)
	velocity := state.GetVelocity()
	if velocity == nil {
		t.Fatalf("expected velocity to remain populated")
	}
	//3.- Confirm the resulting speed is capped by the shared configuration.
	speed := math.Sqrt(velocity.X*velocity.X + velocity.Y*velocity.Y + velocity.Z*velocity.Z)
	if math.Abs(speed-stats.MaxSpeedMps) > 1e-6 {
		t.Fatalf("linear clamp mismatch: got %.6f want %.6f", speed, stats.MaxSpeedMps)
	}
	displacement := state.GetPosition()
	if displacement == nil {
		t.Fatalf("expected position after integration")
	}
	//4.- Ensure the positional delta aligns with the clamped velocity magnitude.
	dispSq := displacement.X*displacement.X + displacement.Y*displacement.Y + displacement.Z*displacement.Z
	if math.Abs(dispSq-stats.MaxSpeedMps*stats.MaxSpeedMps) > 1e-3 {
		t.Fatalf("position delta mismatch: got %.6f want %.6f", dispSq, stats.MaxSpeedMps*stats.MaxSpeedMps)
	}
	angular := state.GetAngularVelocity()
	if angular == nil {
		t.Fatalf("expected angular velocity to remain populated")
	}
	//5.- Validate the angular magnitude matches the configuration limit as well.
	angularSpeed := math.Sqrt(angular.X*angular.X + angular.Y*angular.Y + angular.Z*angular.Z)
	if math.Abs(angularSpeed-stats.MaxAngularSpeedDegPerSec) > 1e-6 {
		t.Fatalf("angular clamp mismatch: got %.6f want %.6f", angularSpeed, stats.MaxAngularSpeedDegPerSec)
	}
}

File: typescript-client/src/physics/integrator.test.ts
import assert from "node:assert";
import { deriveStatsWithModifiers, skiffStats } from "../gameplayConfig";
import {
  GuidanceSpline,
  applyAssistAlignment,
  integrateVehicle,
  wrapAngleDeg,
  type VehicleStateLike,
} from "./integrator";

//1.- Verify the integrator advances both linear and angular state over a step.
{
  const state: VehicleStateLike = {
    position: { x: 1, y: 2, z: 3 },
    velocity: { x: 4, y: -2, z: 0.5 },
    orientation: { yawDeg: 10, pitchDeg: -5, rollDeg: 0 },
    angularVelocity: { x: 20, y: 30, z: -10 },
  };
  integrateVehicle(state, 0.5, skiffStats);
  assert.ok(Math.abs((state.position?.x ?? 0) - 3) < 1e-9, "unexpected x");
  assert.ok(Math.abs((state.position?.y ?? 0) - 1) < 1e-9, "unexpected y");
  assert.ok(Math.abs((state.position?.z ?? 0) - 3.25) < 1e-9, "unexpected z");
  assert.ok(Math.abs((state.orientation?.yawDeg ?? 0) - 25) < 1e-9, "unexpected yaw");
  assert.ok(Math.abs((state.orientation?.pitchDeg ?? 0) - 5) < 1e-9, "unexpected pitch");
  assert.ok(Math.abs((state.orientation?.rollDeg ?? 0) + 5) < 1e-9, "unexpected roll");
}

//2.- Confirm wrapAngleDeg keeps values within the inclusive-exclusive range.
{
  assert.strictEqual(wrapAngleDeg(540), -180 + 0, "540 should wrap to -180");
  assert.ok(wrapAngleDeg(-725) >= -180 && wrapAngleDeg(-725) < 180, "wrapped range");
}

//3.- Ensure assist alignment reorients the craft along the spline.
{
  const spline = new GuidanceSpline([
    { x: 0, y: 0, z: 0 },
    { x: 0, y: 5, z: 5 },
  ]);
  const state: VehicleStateLike = {
    position: { x: 0, y: 0, z: 0 },
    orientation: { yawDeg: -90, pitchDeg: 0, rollDeg: 30 },
    angularVelocity: { x: 15, y: -40, z: 5 },
    flightAssistEnabled: true,
  };
  applyAssistAlignment(state, spline);
  assert.ok(Math.abs(state.orientation?.yawDeg ?? 0) < 1e-9, "yaw alignment");
  assert.ok(Math.abs((state.orientation?.pitchDeg ?? 0) - 45) < 1e-9, "pitch alignment");
  assert.ok(Math.abs(state.orientation?.rollDeg ?? 0) < 1e-9, "roll alignment");
  assert.strictEqual(state.angularVelocity?.x, 0, "angular x dampened");
  assert.strictEqual(state.angularVelocity?.y, 0, "angular y dampened");
  assert.strictEqual(state.angularVelocity?.z, 0, "angular z dampened");
}

//4.- Validate velocity and rotation are clamped by the shared Skiff stats.
{
  const state: VehicleStateLike = {
    position: { x: 0, y: 0, z: 0 },
    velocity: {
      x: skiffStats.maxSpeedMps * 5,
      y: skiffStats.maxSpeedMps * 2,
      z: skiffStats.maxSpeedMps * 0.5,
    },
    orientation: { yawDeg: 0, pitchDeg: 0, rollDeg: 0 },
    angularVelocity: {
      x: skiffStats.maxAngularSpeedDegPerSec * 4,
      y: skiffStats.maxAngularSpeedDegPerSec * 2,
      z: 0,
    },
  };
  integrateVehicle(state, 1, skiffStats);
  const velocity = state.velocity!;
  const speed = Math.sqrt(velocity.x ** 2 + velocity.y ** 2 + velocity.z ** 2);
  assert.ok(Math.abs(speed - skiffStats.maxSpeedMps) < 1e-6, "linear clamp should match Skiff cap");
  const displacementSq =
    (state.position?.x ?? 0) ** 2 + (state.position?.y ?? 0) ** 2 + (state.position?.z ?? 0) ** 2;
  assert.ok(Math.abs(displacementSq - skiffStats.maxSpeedMps ** 2) < 1e-4, "position delta reflects clamped speed");
  const angular = state.angularVelocity!;
  const angularSpeed = Math.sqrt(angular.x ** 2 + angular.y ** 2 + angular.z ** 2);
  assert.ok(
    Math.abs(angularSpeed - skiffStats.maxAngularSpeedDegPerSec) < 1e-6,
    "angular clamp should match Skiff cap",
  );
}

//5.- Loadout specific stats should influence the clamp thresholds on demand.
{
  const aggressiveStats = deriveStatsWithModifiers(skiffStats, {
    speedMultiplier: 1.2,
    agilityMultiplier: 0.5,
    damageMultiplier: 1.0,
    boostCooldownScale: 1.0,
  });
  const state: VehicleStateLike = {
    position: { x: 0, y: 0, z: 0 },
    velocity: { x: aggressiveStats.maxSpeedMps * 10, y: 0, z: 0 },
    orientation: { yawDeg: 0, pitchDeg: 0, rollDeg: 0 },
    angularVelocity: { x: aggressiveStats.maxAngularSpeedDegPerSec * 4, y: 0, z: 0 },
  };
  integrateVehicle(state, 1, aggressiveStats);
  assert.ok(
    Math.abs(Math.sqrt(state.velocity!.x ** 2) - aggressiveStats.maxSpeedMps) < 1e-6,
    "custom stats should clamp linear speed",
  );
  assert.ok(
    Math.abs(Math.sqrt(state.angularVelocity!.x ** 2) - aggressiveStats.maxAngularSpeedDegPerSec) < 1e-6,
    "custom stats should clamp angular speed",
  );
}
